src/Common/DateLUTImpl.h

```diff
@@ -895,6 +895,19 @@ class DateLUTImpl
         return toRelativeHourNum(lut[toLUTIndex(v)].date);
     }
 
+    /// The same formula is used for positive time (after Unix epoch) and negative time (before Unix epoch).
```

Comment:
Why exactly can `toRelativeHourNum` not be used from "date_diff"?

Reply:
@rschu1ze Consider the function:

```
inline Time toRelativeHourNum(Time t) const
{
    if (t >= 0 && offset_is_whole_number_of_hours_during_epoch)
        return t / 3600;

    /// Assume that if offset was fractional, then the fraction is the same as at the beginning of epoch.
    /// NOTE This assumption is false for "Pacific/Pitcairn" and "Pacific/Kiritimati" time zones.
    return (t + DATE_LUT_ADD + 86400 - offset_at_start_of_epoch) / 3600 - (DATE_LUT_ADD / 3600);
}

```

For positive values (dates after Unix epoch) we use the algorithm:

```
if (t >= 0 && offset_is_whole_number_of_hours_during_epoch)
    return t / 3600;

```    

For negative values (dates before Unix epoch) we use the algorithm:

`return (t + DATE_LUT_ADD + 86400 - offset_at_start_of_epoch) / 3600 - (DATE_LUT_ADD / 3600);
`
This way it works fine with both arguments before or both after Unix epoch.

But if start date is before Unix epoch and end date is after Unix epoch then it produces a bias mainly because of using 86400 - offset_at_start_of_epoch for dates before Unix epoch. So it does not work, see #41978

That why in toStableRelativeHourNum we need to use the same formula:

`return (t + DATE_LUT_ADD + 86400 - offset_at_start_of_epoch) / 3600 - (DATE_LUT_ADD / 3600);
`
to calculate relative dates before and after Unix epoch. 

Consider the function:

```
inline Time toRelativeMinuteNum(Time t) const /// NOLINT
{
    return (t + DATE_LUT_ADD) / 60 - (DATE_LUT_ADD / 60);
}

```    
So it uses the same formula for negative and positive dates.
    
But the "big" question is why toRelativeHourNum is "time-zone aware" while toRelativeMinuteNum does not know about time zones? Why what we do for **hours** we don't do for **minutes**?

Reply:
I also don't understand the meaning of the calculation:

`return (t + DATE_LUT_ADD) / 60 - (DATE_LUT_ADD / 60);`

Because the result is about the same as:

`return t / 60;`

If we want to make result type non-negative then we need to do just:
    
`return (t + DATE_LUT_ADD) / 60;`

But `toRelativeMinuteNum` returns `Time` which is signed `Int64`. So it's OK to have negative times before Unix epoch.

Reply:
BTW I found the following issues which are reproducted in current master (should we report it as a separate issue? or probably fix it in current PR?)

It's related when start date is Unix epoch:
```

SELECT dateDiff('minute', toDate('1970-01-01'), toDateTime('1970-01-02 00:00:00'));
-4294965856 - wrong, must be 1440

SELECT dateDiff('minute', toDateTime('1970-01-02 00:00:00'), toDate('1970-01-01'));
4294965856 - wrong, must be -1440

SELECT dateDiff('minute', toDate('1970-01-01'), toDateTime('1970-01-03 00:00:00'));
-4294964416 - wrong, must be 2880

SELECT dateDiff('minute', toDate('1970-01-02'), toDateTime('1970-01-03 00:00:00'));
1440 - OK

SELECT dateDiff('minute', toDate('1970-01-03'), toDateTime('1970-01-04 00:00:00'));
1440 - OK
```

Reply:
Probably we need to create new function:
toStableRelativeMinuteNum
and make it time zone "aware":

```
inline Time toStableRelativeMinuteNum(Time t) const /// NOLINT
{
    return (t + DATE_LUT_ADD + 86400 - offset_at_start_of_epoch) / 60 - (DATE_LUT_ADD / 60);
}
```
So it will be used in dateDiff only, the same way as toStableRelativeHourNum now. 
And probably it will fix the issue above.

Reply:
BTW the issue is re-produced if non-UTC timezone is specified in config.xml, for example:
<timezone>Europe/Paris</timezone>

But if time-zone is not defined or it's defined as UTC
<timezone>UTC</timezone>

Then it works OK:
```
SELECT dateDiff('minute', toDate('1970-01-01'), toDateTime('1970-01-02 00:00:00'));
1440
```
Maybe it's because Date is not time-zone aware and DateTime is timezone aware.

Reply:
I read the earlier discussion (https://github.com/ClickHouse/ClickHouse/pull/42308#issuecomment-1282662137, https://github.com/ClickHouse/ClickHouse/pull/42308#issuecomment-1282749612). I think I misunderstood what was meant by "constraint".

The [documentation](https://clickhouse.com/docs/en/sql-reference/functions/date-time-functions/#date_diff) describes that `date_diff` has the intuitive behavior, i.e. how many sec/min/h/d/w/m/q/y-s do I need to add or subtract to get from one date/time to the other. The function just happens to be implemented using `toRelativeUnitNum()` because that is convenient from a programming POV. More specifically, there is *no* obligation/constraint to replicate (overflow) bugs in `toRelativeUnitNum()` into `date_diff()`, in fact I would consider bug-by-bug-compatibility a bug in itself 😄

Reply:
My understanding is that this PR technically breaks backward compact but only in a quite obscure edge case, namely for `date_diff('hour', A, B)` where A and B are 'DateTime64' and smaller and bigger than 1970-01-01 respectively. Besides that, it adds support for `Date32` arguments to `date_diff` but that cannot be considered backward incompatible.